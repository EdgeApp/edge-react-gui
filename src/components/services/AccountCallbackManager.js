// @flow

import { type EdgeAccount, type EdgeCurrencyWallet } from 'edge-core-js'
import { watchSecurityAlerts } from 'edge-login-ui-rn'
import * as React from 'react'

import { updateExchangeRates } from '../../actions/ExchangeRateActions.js'
import { checkPasswordRecovery } from '../../actions/RecoveryReminderActions.js'
import { newTransactionsRequest, refreshTransactionsRequest } from '../../actions/TransactionListActions.js'
import { updateWalletLoadingProgress, updateWalletsRequest } from '../../actions/WalletActions.js'
import { useAsyncEffect } from '../../hooks/useAsyncEffect.js'
import { useWalletsSubscriber } from '../../hooks/useWalletsSubscriber.js'
import { useEffect, useState } from '../../types/reactHooks.js'
import { useDispatch } from '../../types/reactRedux.js'
import { type NavigationProp, useNavigation } from '../../types/routerTypes.js'
import { isReceivedTransaction, snooze } from '../../util/utils.js'
import { WcSmartContractModal } from '../modals/WcSmartContractModal.js'
import { Airship } from './AirshipInstance.js'

type Props = {
  account: EdgeAccount
}

// Tracks items that need a refresh:
type DirtyList = {
  rates: boolean,
  walletList: boolean,
  wallets: { [walletId: string]: EdgeCurrencyWallet }
}

const notDirty: DirtyList = {
  rates: false,
  walletList: false,
  wallets: {}
}

export function AccountCallbackManager(props: Props) {
  const { account } = props
  const dispatch = useDispatch()
  const [dirty, setDirty] = useState<DirtyList>(notDirty)
  const navigation: NavigationProp<'edge'> = useNavigation()

  // Helper for marking wallets dirty:
  function addWallet(wallet: EdgeCurrencyWallet) {
    setDirty(dirty => ({
      ...dirty,
      rates: true, // We might reconsider this behavior
      wallets: { ...dirty.wallets, [wallet.id]: wallet }
    }))
  }

  // Subscribe to the account:
  useEffect(() => {
    const cleanups = [
      account.watch('currencyWallets', () =>
        setDirty(dirty => ({
          ...dirty,
          walletList: true
        }))
      ),

      account.watch('loggedIn', () => {
        if (account.loggedIn === false) {
          Airship.clear()
          console.log('onLoggedOut')
        }
      }),

      watchSecurityAlerts(account, hasAlerts => {
        navigation.push('securityAlerts')
      }),

      account.rateCache.on('update', () =>
        setDirty(dirty => ({
          ...dirty,
          rates: true
        }))
      )
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  }, [account])

  // Subscribe to each wallet that comes online:
  useWalletsSubscriber(account, wallet => {
    const cleanups = [
      wallet.watch('syncRatio', ratio => {
        dispatch(updateWalletLoadingProgress(wallet.id, ratio))
      }),

      wallet.on('newTransactions', transactions => {
        console.log(`${walletPrefix(wallet)}: onNewTransactions: ${transactions.map(tx => tx.txid).join(' ')}`)

        dispatch(refreshTransactionsRequest(wallet.id, transactions))
        dispatch(newTransactionsRequest(wallet.id, transactions))
        addWallet(wallet)

        // Check if password recovery is set up:
        const finalTxIndex = transactions.length - 1
        if (isReceivedTransaction(transactions[finalTxIndex])) {
          dispatch(checkPasswordRecovery())
        }
      }),

      wallet.on('transactionsChanged', transactions => {
        console.log(`${walletPrefix(wallet)}: onTransactionsChanged: ${transactions.map(tx => tx.txid).join(' ')}`)

        dispatch(refreshTransactionsRequest(wallet.id, transactions))
        addWallet(wallet)
      }),

      wallet.on('wcNewContractCall', obj => {
        const { dApp, payload, uri, walletId } = obj
        if (walletId == null) return
        Airship.show(bridge => <WcSmartContractModal bridge={bridge} walletId={walletId} dApp={dApp} payload={payload} uri={uri} />)
      }),

      // These ones defer their work until later:
      wallet.watch('balances', () => addWallet(wallet)),
      wallet.watch('enabledTokenIds', () => addWallet(wallet)),
      wallet.watch('name', () => addWallet(wallet))
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  })

  // Do the expensive work with rate limiting:
  useAsyncEffect(async () => {
    setDirty(notDirty)

    // Update wallets:
    const walletIds = Object.keys(dirty.wallets)
    if (dirty.walletList) {
      // Update all wallets (hammer mode):
      console.log('Updating wallet list')
      await dispatch(updateWalletsRequest())
      await snooze(1000)
    } else if (walletIds.length > 0) {
      // Update individual wallets:
      console.log(`Updating wallets: ${walletIds.map(id => walletPrefix(dirty.wallets[id])).join(' ')}`)
      dispatch({
        type: 'UI/WALLETS/UPSERT_WALLETS',
        data: {
          wallets: walletIds.map(id => dirty.wallets[id])
        }
      })
      await snooze(1000)
    }

    // Update exchange rates:
    if (dirty.rates) {
      console.log('Updating exchange rates')
      await dispatch(updateExchangeRates())
      await snooze(1000)
    }
  }, [dirty])

  return null
}

function walletPrefix(wallet: EdgeCurrencyWallet): string {
  return `${wallet.currencyInfo.currencyCode}-${wallet.id.slice(0, 2)}`
}
