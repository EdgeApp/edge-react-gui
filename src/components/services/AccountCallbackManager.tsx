import { EdgeAccount, EdgeCurrencyWallet } from 'edge-core-js'
import { watchSecurityAlerts } from 'edge-login-ui-rn'
import * as React from 'react'

import { showBackupModal } from '../../actions/BackupModalActions'
import { updateExchangeRates } from '../../actions/ExchangeRateActions'
import { checkFioObtData } from '../../actions/FioActions'
import { showReceiveDropdown } from '../../actions/ReceiveDropdown'
import { checkPasswordRecovery } from '../../actions/RecoveryReminderActions'
import { updateWalletLoadingProgress, updateWalletsRequest } from '../../actions/WalletActions'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useWalletsSubscriber } from '../../hooks/useWalletsSubscriber'
import { stakeMetadataCache } from '../../plugins/stake-plugins/metadataCache'
import { useDispatch } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { snooze } from '../../util/utils'
import { Airship } from './AirshipInstance'

interface Props {
  account: EdgeAccount
  navigation: NavigationBase
}

// Tracks items that need a refresh:
interface DirtyList {
  rates: boolean
  walletList: boolean
}

const notDirty: DirtyList = {
  rates: false,
  walletList: false
}

export function AccountCallbackManager(props: Props) {
  const { account, navigation } = props
  const dispatch = useDispatch()
  const [dirty, setDirty] = React.useState<DirtyList>(notDirty)

  // Helper for marking wallets dirty:
  function setRatesDirty() {
    setDirty(dirty => ({
      ...dirty,
      rates: true
    }))
  }

  // Subscribe to the account:
  React.useEffect(() => {
    const cleanups = [
      account.watch('currencyWallets', () =>
        setDirty(dirty => ({
          ...dirty,
          walletList: true
        }))
      ),

      account.watch('loggedIn', () => {
        if (!account.loggedIn) {
          Airship.clear()
          console.log('onLoggedOut')
        }
      }),

      watchSecurityAlerts(account, hasAlerts => {
        if (hasAlerts) {
          navigation.navigate('securityAlerts', {})
        }
      }),

      account.rateCache.on('update', () =>
        setDirty(dirty => ({
          ...dirty,
          rates: true
        }))
      )
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  }, [account, navigation])

  // Subscribe to each wallet that comes online:
  useWalletsSubscriber(account, wallet => {
    const cleanups = [
      wallet.watch('syncRatio', ratio => {
        dispatch(updateWalletLoadingProgress(wallet.id, ratio))
      }),

      wallet.on('newTransactions', transactions => {
        for (const tx of transactions) {
          const txid = tx.txid.toLowerCase()
          const cacheEntries = stakeMetadataCache[txid]
          // Assign cached stake metadata
          if (cacheEntries != null) {
            cacheEntries.forEach(cacheEntry => {
              const { currencyCode, metadata } = cacheEntry
              if (tx.currencyCode !== currencyCode) return
              wallet.saveTx({ ...tx, metadata }).catch(err => console.warn(err))
            })

            delete stakeMetadataCache[txid]
          }
        }

        console.log(`${walletPrefix(wallet)}: onNewTransactions: ${transactions.map(tx => tx.txid).join(' ')}`)

        // Check for incoming FIO requests:
        const receivedTxs = transactions.filter(tx => !tx.isSend)
        if (receivedTxs.length > 0) dispatch(checkFioObtData(wallet, receivedTxs)).catch(err => console.warn(err))

        // Show the dropdown for the first transaction:
        const [firstReceive] = receivedTxs
        if (firstReceive != null) {
          dispatch(showReceiveDropdown(navigation, firstReceive))

          // Notify the user to consider backing up their account
          if (account.username == null) {
            showBackupModal({ navigation })
          }
        }

        // Check if password recovery is set up:
        const finalTxIndex = transactions.length - 1
        if (!transactions[finalTxIndex].isSend) {
          dispatch(checkPasswordRecovery(navigation))
        }
      }),

      wallet.on('transactionsChanged', transactions => {
        console.log(`${walletPrefix(wallet)}: onTransactionsChanged: ${transactions.map(tx => tx.txid).join(' ')}`)
      }),

      // These ones defer their work until later:
      wallet.watch('balances', () => setRatesDirty()),
      wallet.watch('enabledTokenIds', () => setRatesDirty())
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  })

  // Do the expensive work with rate limiting:
  useAsyncEffect(
    async () => {
      setDirty(notDirty)

      // Update wallets:
      if (dirty.walletList) {
        // Update all wallets (hammer mode):
        console.log('Updating wallet list')
        await dispatch(updateWalletsRequest())
        await snooze(1000)
      }

      // Update exchange rates:
      if (dirty.rates) {
        console.log('Updating exchange rates')
        await dispatch(updateExchangeRates())
        await snooze(1000)
      }
    },
    [dirty],
    'AccountCallbackManager'
  )

  return null
}

function walletPrefix(wallet: EdgeCurrencyWallet): string {
  return `${wallet.currencyInfo.currencyCode}-${wallet.id.slice(0, 2)}`
}
