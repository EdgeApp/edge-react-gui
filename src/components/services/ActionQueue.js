// @flow

import { type EdgeAccount } from 'edge-core-js'

import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { type ActionQueueMap } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useRef } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueue = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)
  const executingRef = useRef<{ [programId: string]: boolean }>({})

  // TODO: Fix sync server stale ActionQueue state overwrite on startup
  //
  // Initialization
  //

  // useAsyncEffect(async () => {
  //   if (account?.dataStore != null) {
  //     const store = makeActionQueueStore(account)
  //     const queue = await store.getActionQueueMap()
  //     dispatch({
  //       type: 'ACTION_QUEUE/LOAD_QUEUE',
  //       data: queue
  //     })
  //   }
  // }, [account, dispatch])

  //
  // Runtime
  //

  useAsyncEffect(async () => {
    if (queue != null) {
      const executing = executingRef.current
      const promises = Object.keys(queue)
        .filter(
          programId =>
            // Ignore running programs
            !executing[programId]
        )
        .map(async programId => {
          // Set program to running
          executing[programId] = true

          const { program, state } = queue[programId]
          const { nextState } = await executeActionProgram(account, program, state).catch((error: Error) => ({
            nextState: {
              ...state,
              effect: { type: 'done', error }
            }
          }))

          // Update program state
          dispatch(updateActionProgramState(nextState))

          // Unset program to running
          executing[programId] = false
        })

      await Promise.all(promises)
    }
  }, [queue])

  return null
}
