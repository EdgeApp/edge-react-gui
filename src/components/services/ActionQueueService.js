// @flow

import { type EdgeAccount } from 'edge-core-js'

import { makeActionQueueStore } from '../../controllers/action-queue/ActionQueueStore'
import { mockActionProgram } from '../../controllers/action-queue/mock'
import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { type ActionQueueMap, type ExecutionContext } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useRef } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueueService = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const clientId: string = useSelector(state => state.core.context.clientId)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)
  const executingRef = useRef<{ [programId: string]: boolean }>({})

  const executionContext: ExecutionContext = { account, clientId }

  //
  // Initialization
  //

  useAsyncEffect(async () => {
    if (account?.dataStore != null) {
      const store = makeActionQueueStore(account, clientId)
      const queue = await store.getActionQueueMap()
      dispatch({
        type: 'ACTION_QUEUE/LOAD_QUEUE',
        data: queue
      })
    }
  }, [account, dispatch])

  //
  // Runtime
  //

  useAsyncEffect(async () => {
    if (queue != null) {
      const executing = executingRef.current
      const promises = Object.keys(queue)
        .filter(programId => !executing[programId] && queue[programId].state.clientId === clientId)
        .map(async programId => {
          // Set program to running
          executing[programId] = true

          const { program, state } = queue[programId]

          if (program.mockMode) {
            const { nextState } = await mockActionProgram(account, program, state)
            dispatch(updateActionProgramState(nextState))
            executing[programId] = false
            return
          }

          const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
            console.warn(new Error('Action Program Exception: ' + error.message))
            console.error(error)
            return {
              nextState: {
                ...state,
                effect: { type: 'done', error }
              }
            }
          })

          // Update program state
          dispatch(updateActionProgramState(nextState))

          // Unset program to running
          executing[programId] = false
        })

      await Promise.all(promises)
    }
  }, [queue])

  return null
}
