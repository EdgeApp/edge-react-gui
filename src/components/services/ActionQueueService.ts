import { EdgeAccount } from 'edge-core-js'
import * as React from 'react'
import { useRef } from 'react'

import { makeActionQueueStore } from '../../controllers/action-queue/ActionQueueStore'
import { asCleanError } from '../../controllers/action-queue/cleaners'
import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime/executeActionProgram'
import { ActionProgramState, ActionQueueMap, ExecutionResults } from '../../controllers/action-queue/types'
import { wasPushRequestBody } from '../../controllers/action-queue/types/pushApiTypes'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useExecutionContext } from '../../hooks/useExecutionContext'
import { useHandler } from '../../hooks/useHandler'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { makePeriodicTask } from '../../util/PeriodicTask'
import { makePushClient } from '../../util/PushClient/PushClient'

const EXECUTION_INTERVAL = 1000

interface ServiceProgramStates {
  [programId: string]: {
    executing: boolean
  }
}

export const ActionQueueService = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const clientId: string = useSelector(state => state.core.context.clientId)
  const actionQueueMap: ActionQueueMap = useSelector(state => state.actionQueue.actionQueueMap)
  const activeProgramIds = useSelector(state => state.actionQueue.activeProgramIds)
  const serviceProgramStatesRef = useRef<ServiceProgramStates>({})

  const executionContext = useExecutionContext()
  const executionContextMock = useExecutionContext(true)

  const updateProgramState = useHandler(async (state: ActionProgramState, executing: boolean) => {
    const { programId } = state
    // Update service level state
    const serviceProgramStates = serviceProgramStatesRef.current
    serviceProgramStates[programId] = { ...serviceProgramStates[programId], executing }
    // Update action queue level state
    await dispatch(
      updateActionProgramState({
        ...state,
        executing: executing
      })
    )
  })

  //
  // Log Dev Info
  //
  React.useEffect(() => {
    if (__DEV__ && account != null && account.rootLoginId != null && clientId != null) {
      const pushClient = makePushClient(account, clientId)
      const requestBody = pushClient.getPushRequestBody()
      console.log('***********************')
      console.log('PUSH SERVER DEV INFO:', wasPushRequestBody(requestBody))
      console.log('***********************')
    }
  }, [account, clientId])

  //
  // Initialization
  //

  useAsyncEffect(
    async () => {
      if (account?.dataStore != null) {
        const store = makeActionQueueStore(account, clientId)
        const queue = await store.getActionQueueMap()
        dispatch({
          type: 'ACTION_QUEUE/LOAD_QUEUE',
          data: queue
        })
      }
    },
    [account, dispatch],
    'ActionQueueService'
  )

  //
  // Runtime Loop
  //

  React.useEffect(() => {
    const serviceProgramStates = serviceProgramStatesRef.current

    // Loop function
    const task = async () => {
      // Programs that require immediate attention
      const urgentProgramIds = activeProgramIds.filter(programId => {
        const { state } = actionQueueMap[programId]
        // Don't execute programs which are assigned to another client.
        // This predicate must come first to avoid the device/client from
        // impacting program heing handled by other clients.
        if (state.clientId !== clientId) return false
        // Don't execute programs which are currently being executed by service
        if (state.executing && serviceProgramStates[programId]) {
          return false
        }
        // Detect program interruption if service is not evaluating the program
        // and program evaluation was executing an action op.
        if (state.executing && !serviceProgramStates[programId] && state.effective) {
          dispatch(
            updateActionProgramState({
              ...state,
              effect: { type: 'done', error: new Error('Program Interrupted') },
              executing: false
            })
          )
            // Use console.warn to not spam user
            .catch(err => console.warn(err))
          return false
        }
        // Don't execute programs which have not reached their scheduled time
        if (!state.effective && state.nextExecutionTime > Date.now()) return false
        // Otherwise, it's safe to execute
        return true
      })
      // Act on urgent programs
      const promises = urgentProgramIds.map(async programId => {
        const { program, state } = actionQueueMap[programId]

        // Set program state to executing
        await updateProgramState(state, true)

        // Use mock execution function if program is marked as mockMode
        const context = program.mockMode ? executionContextMock : executionContext

        const { nextState } = await executeActionProgram(context, program, state).catch((error): ExecutionResults => {
          const cleanError = asCleanError(error)
          console.warn(new Error('Action Program Exception: ' + cleanError.message))
          console.error(cleanError)
          return {
            nextState: {
              ...state,
              effect: { type: 'done', error: cleanError },
              effective: true,
              nextExecutionTime: -1
            }
          }
        })

        // Update program state
        await updateProgramState(nextState, false)
      })
      await Promise.all(promises)
    }

    // Setup loop
    const periodicTask = makePeriodicTask(task, EXECUTION_INTERVAL, {
      onError: error => {
        const errorMessage = String(error)
        console.warn(new Error('Unexpected error in ActionQueueService: ' + errorMessage))
        console.error(error)
      }
    })

    periodicTask.start()

    // Cleanup loop
    return () => periodicTask.stop()
  }, [account, clientId, dispatch, updateProgramState, actionQueueMap, activeProgramIds, executionContextMock, executionContext])

  // Return no component/view
  return null
}
