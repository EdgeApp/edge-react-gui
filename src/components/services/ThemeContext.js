// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
