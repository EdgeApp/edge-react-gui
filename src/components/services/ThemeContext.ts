import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
