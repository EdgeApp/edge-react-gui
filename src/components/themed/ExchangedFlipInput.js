// @flow

import { div, log10, mul } from 'biggystring'
import { hook } from 'cavy'
import * as React from 'react'

import type { GuiCurrencyInfo } from '../../types/types.js'
import { DECIMAL_PRECISION, getObjectDiff, precisionAdjust, zeroString } from '../../util/utils.js'
import type { FlipInputFieldInfo } from './FlipInput.js'
import { FlipInput } from './FlipInput.js'

export type ExchangedFlipInputAmounts = {
  exchangeAmount: string,
  nativeAmount: string
}

export type ExchangedFlipInputOwnProps = {
  // Initial amount of the primary field in `exchangeAmount` denomination. This is converted to a `decimalAmount`
  // in the proper display denomination to be passed into FlipInput
  overridePrimaryExchangeAmount: string,
  primaryCurrencyInfo: GuiCurrencyInfo,
  secondaryCurrencyInfo: GuiCurrencyInfo,
  onNext?: () => void,
  onFocus?: () => void,
  onBlur?: () => void,

  // Exchange rate
  exchangeSecondaryToPrimaryRatio: string,

  forceUpdateGuiCounter: number,
  keyboardVisible: boolean,

  // Callback for when the `primaryAmount` changes. This returns both a `nativeAmount` and an `exchangeAmount`. Both
  // amounts are ONLY for the primary field. Parent will not be given values for the secondary field.
  onExchangeAmountChanged(amounts: ExchangedFlipInputAmounts): mixed,
  isEditable: boolean,
  isFiatOnTop: boolean,
  isFocus: boolean,

  topReturnKeyType?: string,
  inputAccessoryViewID?: string,
  headerText: string,
  headerCallback?: () => void,
  onError?: (error: string | void) => void
}

type Props = ExchangedFlipInputOwnProps

type State = {
  overridePrimaryDecimalAmount: string, // This should be a decimal amount in display denomination (ie. mBTC)
  exchangeSecondaryToPrimaryRatio: string,
  primaryInfo: FlipInputFieldInfo,
  secondaryInfo: FlipInputFieldInfo
}

function getPrimaryDisplayToExchangeRatio(props: Props): string {
  const exchangeMultiplier: string = props.primaryCurrencyInfo.exchangeDenomination.multiplier
  const displayMultiplier: string = props.primaryCurrencyInfo.displayDenomination.multiplier
  return div(exchangeMultiplier, displayMultiplier, DECIMAL_PRECISION)
}

function getSecondaryDisplayToExchangeRatio(props: Props): string {
  const displayMultiplier: string = props.secondaryCurrencyInfo.displayDenomination.multiplier
  const exchangeMultiplier: string = props.secondaryCurrencyInfo.exchangeDenomination.multiplier
  return div(exchangeMultiplier, displayMultiplier, DECIMAL_PRECISION)
}

function propsToState(props: Props): State {
  // Calculate secondaryToPrimaryRatio for FlipInput. FlipInput takes a ratio in display amounts which may be
  // different than exchange amounts. ie. USD / mBTC
  // nextProps.exchangeSecondaryToPrimaryRatio // ie. 1/10000
  const primaryDisplayToExchangeRatio = getPrimaryDisplayToExchangeRatio(props) // 1/1000 for mBTC
  const secondaryDisplayToExchangeRatio = getSecondaryDisplayToExchangeRatio(props) // 1 for USD
  let exchangeSecondaryToPrimaryRatio = div(props.exchangeSecondaryToPrimaryRatio, primaryDisplayToExchangeRatio, DECIMAL_PRECISION) // Should be 1/10

  exchangeSecondaryToPrimaryRatio = mul(exchangeSecondaryToPrimaryRatio, secondaryDisplayToExchangeRatio) // Noop usually for USD since we only ever use the same exchange and display multiplier

  // Calculate FlipInputFieldInfo from GuiCurrencyInfos
  const secondaryPrecision: number = log10(props.secondaryCurrencyInfo.displayDenomination.multiplier)
  const primaryEntryPrecision = log10(props.primaryCurrencyInfo.displayDenomination.multiplier)
  // Limit the precision of the primaryPrecision by what would be no more
  // than 0.01 (of whateve fiat currency) accuracy when converting a fiat value into a crypto value.
  //
  // Assume secondaryInfo refers to a fiatValue and take the secondaryToPrimaryRatio (exchange rate) and
  // see how much precision this crypto denomination needs to achieve accuracy to 0.01 units of the current fiat
  // currency. To do this we need to compare the "exchangeDenomination" of primaryInfo and secondaryInfo since
  // only those values are relevant to secondaryToPrimaryRatio
  const precisionAdjustVal = precisionAdjust({
    primaryExchangeMultiplier: props.primaryCurrencyInfo.exchangeDenomination.multiplier,
    secondaryExchangeMultiplier: props.secondaryCurrencyInfo.exchangeDenomination.multiplier,
    exchangeSecondaryToPrimaryRatio: props.exchangeSecondaryToPrimaryRatio
  })

  const newPrimaryPrecision = primaryEntryPrecision - precisionAdjustVal
  const primaryConversionPrecision = newPrimaryPrecision >= 0 ? newPrimaryPrecision : 0

  const primaryInfo: FlipInputFieldInfo = {
    currencyName: props.primaryCurrencyInfo.displayDenomination.name,
    currencySymbol: props.primaryCurrencyInfo.displayDenomination.symbol ? props.primaryCurrencyInfo.displayDenomination.symbol : '',
    currencyCode: props.primaryCurrencyInfo.displayCurrencyCode,
    pluginId: props.primaryCurrencyInfo.pluginId,
    maxEntryDecimals: primaryEntryPrecision,
    maxConversionDecimals: primaryConversionPrecision
  }

  const secondaryInfo: FlipInputFieldInfo = {
    currencyName: props.secondaryCurrencyInfo.displayDenomination.name,
    currencySymbol: props.secondaryCurrencyInfo.displayDenomination.symbol ? props.secondaryCurrencyInfo.displayDenomination.symbol : '',
    currencyCode: props.secondaryCurrencyInfo.displayCurrencyCode,
    maxEntryDecimals: secondaryPrecision,
    maxConversionDecimals: secondaryPrecision
  }

  // Convert overridePrimaryExchangeAmount => overridePrimaryDecimalAmount which goes from exchange to display
  // ie BTC to mBTC
  const overridePrimaryDecimalAmount = mul(props.overridePrimaryExchangeAmount, primaryDisplayToExchangeRatio)

  return { primaryInfo, secondaryInfo, exchangeSecondaryToPrimaryRatio, overridePrimaryDecimalAmount }
}

export class ExchangedFlipInputComponent extends React.Component<Props, State> {
  flipInput: React.ElementRef<typeof FlipInput> | null = null

  static defaultProps = {
    isEditable: true
  }

  constructor(props: Props) {
    super(props)
    this.state = propsToState(props)
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    this.setState(propsToState(nextProps))
  }

  shouldComponentUpdate(nextProps: Props, nextState: State) {
    let diffElement2: string = ''
    const diffElement = getObjectDiff(this.props, nextProps, {
      primaryCurrencyInfo: true,
      secondaryCurrencyInfo: true
    })
    if (!diffElement) {
      diffElement2 = getObjectDiff(this.state, nextState, {
        primaryInfo: true,
        secondaryInfo: true
      })
    }
    return !!diffElement || !!diffElement2
  }

  onAmountChanged = (decimalAmount: string): void => {
    const exchangeAmount = div(decimalAmount, getPrimaryDisplayToExchangeRatio(this.props), DECIMAL_PRECISION)
    const nativeAmount = mul(exchangeAmount, this.props.primaryCurrencyInfo.exchangeDenomination.multiplier)
    this.props.onExchangeAmountChanged({ exchangeAmount, nativeAmount })
  }

  isFiatOnTop = () => {
    if (!this.props.isFiatOnTop) {
      return false
    }
    return !zeroString(this.state.exchangeSecondaryToPrimaryRatio)
  }

  toggleCryptoOnBottom = () => {
    if (this.flipInput != null) {
      this.flipInput.toggleCryptoOnBottom()
    }
  }

  render() {
    return (
      <FlipInput
        overridePrimaryDecimalAmount={this.state.overridePrimaryDecimalAmount}
        exchangeSecondaryToPrimaryRatio={this.state.exchangeSecondaryToPrimaryRatio}
        headerText={this.props.headerText}
        headerCallback={this.props.headerCallback}
        primaryInfo={this.state.primaryInfo}
        secondaryInfo={this.state.secondaryInfo}
        forceUpdateGuiCounter={this.props.forceUpdateGuiCounter}
        onAmountChanged={this.onAmountChanged}
        onError={this.props.onError}
        keyboardVisible={this.props.keyboardVisible}
        isEditable={this.props.isEditable}
        isFiatOnTop={this.isFiatOnTop()}
        isFocus={this.props.isFocus}
        onNext={this.props.onNext}
        onFocus={this.props.onFocus}
        onBlur={this.props.onBlur}
        topReturnKeyType={this.props.topReturnKeyType}
        inputAccessoryViewID={this.props.inputAccessoryViewID}
        ref={ref => (this.flipInput = ref)}
      />
    )
  }
}
export const ExchangedFlipInput = hook(ExchangedFlipInputComponent)
