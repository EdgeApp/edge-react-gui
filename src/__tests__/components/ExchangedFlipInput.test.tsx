/* globals describe it expect */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { getTheme } from '../../components/services/ThemeContext'
import { ExchangedFlipInput } from '../../components/themed/ExchangedFlipInput'

describe('ExchangedFlipInput', () => {
  it('should render with loading props', () => {
    // @ts-expect-error
    const renderer = new ShallowRenderer()

    const props = {
      overridePrimaryExchangeAmount: 'string',
      primaryCurrencyInfo: {
        displayCurrencyCode: 'BTC',
        exchangeCurrencyCode: 'BTC',
        displayDenomination: {
          multiplier: '10000000000',
          name: 'BTC',
          symbol: '₿'
        },
        exchangeDenomination: {
          multiplier: '10000000000',
          name: 'Bitcoin',
          symbol: '₿'
        }
      },
      secondaryCurrencyInfo: {
        displayCurrencyCode: 'BTC',
        exchangeCurrencyCode: 'BTC',
        displayDenomination: {
          multiplier: '10000000000',
          name: 'Bitcoin',
          symbol: '₿'
        },
        exchangeDenomination: {
          multiplier: '10000000000',
          name: 'Bitcoin',
          symbol: '₿'
        }
      },
      exchangeSecondaryToPrimaryRatio: '0',
      forceUpdateGuiCounter: 11,
      keyboardVisible: true,
      // @ts-expect-error
      onExchangeAmountChanged: amounts => ['123123', '123123', 'BTC'],
      isEditable: true,
      isFiatOnTop: true,
      isFocus: true,
      headerText: '123',
      theme: getTheme()
    }
    // @ts-expect-error
    const actual = renderer.render(<ExchangedFlipInput {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
