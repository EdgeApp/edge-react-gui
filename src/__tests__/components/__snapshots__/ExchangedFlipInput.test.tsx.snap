// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangedFlipInput should render with loading props 1`] = `
<ForwardRef
  exchangeSecondaryToPrimaryRatio="0"
  forceUpdateGuiCounter={0}
  headerText="123"
  isEditable={true}
  isFiatOnTop={false}
  isFocus={true}
  keyboardVisible={true}
  onAmountChanged={[Function]}
  overridePrimaryDecimalAmount="3119046"
  primaryInfo={
    {
      "currencyCode": "BTC",
      "currencyName": "BTC",
      "currencySymbol": "₿",
      "maxConversionDecimals": 0,
      "maxEntryDecimals": 10,
      "pluginId": undefined,
    }
  }
  secondaryInfo={
    {
      "currencyCode": "BTC",
      "currencyName": "BTC",
      "currencySymbol": "₿",
      "maxConversionDecimals": 10,
      "maxEntryDecimals": 10,
    }
  }
/>
`;
